<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="author" title="About these documents" href="../about.html" /><link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="How to index GO-CAMs with LinkML-Store" href="Index-GO-CAMs.html" /><link rel="prev" title="Example: Storing an ontology" href="Store-an-Ontology.html" />

    <link rel="shortcut icon" href="https://linkml.io/uploads/linkml-logo_color-no-words.png"/><!-- Generated with Sphinx 7.4.7 and Furo 2024.08.06 -->
        <title>How to index Phenopackets with LinkML-Store - linkml-store documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=fd3f3429" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">linkml-store  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="https://linkml.io/uploads/linkml-logo_color.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">linkml-store  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about.html">About</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../tutorials/index.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/Command-Line-Tutorial.html">Tutorial: Using the Command Line Interface</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/Python-Tutorial.html">Tutorial: Using the Python API</a></li>
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="index.html">How-To Guides</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of How-To Guides</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Use-MongoDB.html">How to use MongoDB with LinkML-Store</a></li>
<li class="toctree-l2"><a class="reference internal" href="Use-Semantic-Search.html">How to use Semantic Search</a></li>
<li class="toctree-l2"><a class="reference internal" href="Query-the-Monarch-KG.html">How to query the Monarch-KG</a></li>
<li class="toctree-l2"><a class="reference internal" href="Query-Solr-using-CLI.html">How to query Solr using the Command Line</a></li>
<li class="toctree-l2"><a class="reference internal" href="Use-Neo4j.html">How to use Neo4J with LinkML-Store</a></li>
<li class="toctree-l2"><a class="reference internal" href="Check-Referential-Integrity.html">How to Check Referential Integrity</a></li>
<li class="toctree-l2"><a class="reference internal" href="Store-an-Ontology.html">Example: Storing an ontology</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">How to index Phenopackets with LinkML-Store</a></li>
<li class="toctree-l2"><a class="reference internal" href="Index-GO-CAMs.html">How to index GO-CAMs with LinkML-Store</a></li>
<li class="toctree-l2"><a class="reference internal" href="Predict-Missing-Data.html">How to predict missing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="Perform-RAG-Inference.html">Perform RAG Inference</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference/index.html">Reference</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference/linkml_store.api.html">API</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../reference/linkml_store.api.stores.html">Adapters</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Adapters</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../reference/linkml_store.api.stores.duckdb.html">linkml_store.api.stores.duckdb package</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of linkml_store.api.stores.duckdb package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.duckdb.duckdb_collection.html">linkml_store.api.stores.duckdb.duckdb_collection module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.duckdb.duckdb_database.html">linkml_store.api.stores.duckdb.duckdb_database module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.duckdb.mappings.html">linkml_store.api.stores.duckdb.mappings module</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../reference/linkml_store.api.stores.mongodb.html">linkml_store.api.stores.mongodb package</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of linkml_store.api.stores.mongodb package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.mongodb.mongodb_collection.html">linkml_store.api.stores.mongodb.mongodb_collection module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.mongodb.mongodb_database.html">linkml_store.api.stores.mongodb.mongodb_database module</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../reference/linkml_store.api.stores.filesystem.html">linkml_store.api.stores.filesystem package</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of linkml_store.api.stores.filesystem package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.filesystem.filesystem_collection.html">linkml_store.api.stores.filesystem.filesystem_collection module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.filesystem.filesystem_database.html">linkml_store.api.stores.filesystem.filesystem_database module</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../reference/linkml_store.api.stores.solr.html">linkml_store.api.stores.solr package</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of linkml_store.api.stores.solr package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.solr.solr_collection.html">linkml_store.api.stores.solr.solr_collection module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.solr.solr_database.html">linkml_store.api.stores.solr.solr_database module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.solr.solr_utils.html">linkml_store.api.stores.solr.solr_utils module</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../reference/linkml_store.api.stores.hdf5.html">linkml_store.api.stores.hdf5 package</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of linkml_store.api.stores.hdf5 package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.hdf5.hdf5_collection.html">linkml_store.api.stores.hdf5.hdf5_collection module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.hdf5.hdf5_database.html">linkml_store.api.stores.hdf5.hdf5_database module</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../reference/linkml_store.api.stores.chromadb.html">linkml_store.api.stores.chromadb package</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of linkml_store.api.stores.chromadb package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.chromadb.chromadb_collection.html">linkml_store.api.stores.chromadb.chromadb_collection module</a></li>
<li class="toctree-l4"><a class="reference internal" href="../reference/linkml_store.api.stores.chromadb.chromadb_database.html">linkml_store.api.stores.chromadb.chromadb_database module</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html#indices-and-tables">Indices and tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/how-to/Index-Phenopackets.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="How-to-index-Phenopackets-with-LinkML-Store">
<h1>How to index Phenopackets with LinkML-Store<a class="headerlink" href="#How-to-index-Phenopackets-with-LinkML-Store" title="Link to this heading"></a></h1>
<section id="Use-pystow-to-download-phenopackets">
<h2>Use pystow to download phenopackets<a class="headerlink" href="#Use-pystow-to-download-phenopackets" title="Link to this heading"></a></h2>
<p>We will download from the Monarch Initiative <a class="reference external" href="https://github.com/monarch-initiative/phenopacket-store">phenopacket-store</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pystow</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">yaml</span>

<span class="n">path</span> <span class="o">=</span> <span class="n">pystow</span><span class="o">.</span><span class="n">ensure_untar</span><span class="p">(</span><span class="s2">&quot;tmp&quot;</span><span class="p">,</span> <span class="s2">&quot;phenopackets&quot;</span><span class="p">,</span> <span class="n">url</span><span class="o">=</span><span class="s2">&quot; https://github.com/monarch-initiative/phenopacket-store/releases/latest/download/all_phenopackets.tgz&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># iterate over all *.json files in the phenopackets directory and parse to an object</span>
<span class="c1"># we will recursively walk the path using os.walk ( we don&#39;t worry about loading yet)</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<span class="n">objs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">root</span><span class="p">,</span> <span class="n">dirs</span><span class="p">,</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.json&quot;</span><span class="p">):</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">file</span><span class="p">))</span> <span class="k">as</span> <span class="n">stream</span><span class="p">:</span>
                <span class="n">obj</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
                <span class="n">objs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">objs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
4876
</pre></div></div>
</div>
</section>
<section id="Creating-a-client-and-attaching-to-a-database">
<h2>Creating a client and attaching to a database<a class="headerlink" href="#Creating-a-client-and-attaching-to-a-database" title="Link to this heading"></a></h2>
<p>First we will create a client as normal:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">linkml_store</span><span class="w"> </span><span class="kn">import</span> <span class="n">Client</span>

<span class="n">client</span> <span class="o">=</span> <span class="n">Client</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Next we’ll attach to a MongoDB instance. this assumes you have one running already.</p>
<p>We will make a database called “phenopackets” and recreate it if it already exists</p>
<p>(note for people running this notebook locally - if you happen to have a database with this name in your current mongo instance it will be deleted!)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">db</span> <span class="o">=</span> <span class="n">client</span><span class="o">.</span><span class="n">attach_database</span><span class="p">(</span><span class="s2">&quot;mongodb://localhost:27017&quot;</span><span class="p">,</span> <span class="s2">&quot;phenopackets&quot;</span><span class="p">,</span> <span class="n">recreate_if_exists</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Creating-a-collection">
<h2>Creating a collection<a class="headerlink" href="#Creating-a-collection" title="Link to this heading"></a></h2>
<p>We’ll create a simple test collection. The concept of collection in linkml-store maps directly to mongodb collections</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">create_collection</span><span class="p">(</span><span class="s2">&quot;main&quot;</span><span class="p">,</span> <span class="n">recreate_if_exists</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Inserting-objects-into-the-store">
<h2>Inserting objects into the store<a class="headerlink" href="#Inserting-objects-into-the-store" title="Link to this heading"></a></h2>
<p>We’ll use the standard <code class="docutils literal notranslate"><span class="pre">insert</span></code> method to insert the phenopackets into the collection. At this stage there is no explicit schema.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">objs</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Check-contents">
<h2>Check contents<a class="headerlink" href="#Check-contents" title="Link to this heading"></a></h2>
<p>We can check the number of rows in the collection, to ensure everything was inserted correctly:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">find</span><span class="p">({},</span> <span class="n">limit</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">num_rows</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
4876
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="k">assert</span> <span class="n">collection</span><span class="o">.</span><span class="n">find</span><span class="p">({},</span> <span class="n">limit</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">num_rows</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">objs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Let’s check with pandas just to make sure it looks as expected; we’ll query for a specific OMIM disease:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">qr</span> <span class="o">=</span> <span class="n">collection</span><span class="o">.</span><span class="n">find</span><span class="p">({</span><span class="s2">&quot;diseases.term.id&quot;</span><span class="p">:</span> <span class="s2">&quot;OMIM:618499&quot;</span><span class="p">},</span> <span class="n">limit</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">qr</span><span class="o">.</span><span class="n">rows_dataframe</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id</th>
      <th>subject</th>
      <th>phenotypicFeatures</th>
      <th>interpretations</th>
      <th>diseases</th>
      <th>metaData</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>PMID_28289718_Higgins-Patient-1</td>
      <td>{'id': 'Higgins-Patient-1', 'timeAtLastEncount...</td>
      <td>[{'type': {'id': 'HP:0001714', 'label': 'Ventr...</td>
      <td>[{'id': 'Higgins-Patient-1', 'progressStatus':...</td>
      <td>[{'term': {'id': 'OMIM:618499', 'label': 'Noon...</td>
      <td>{'created': '2024-03-28T11:11:48.590163946Z', ...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>PMID_31173466_Suzuki-Patient-1</td>
      <td>{'id': 'Suzuki-Patient-1', 'timeAtLastEncounte...</td>
      <td>[{'type': {'id': 'HP:0001714', 'label': 'Ventr...</td>
      <td>[{'id': 'Suzuki-Patient-1', 'progressStatus': ...</td>
      <td>[{'term': {'id': 'OMIM:618499', 'label': 'Noon...</td>
      <td>{'created': '2024-03-28T11:11:48.594725131Z', ...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>PMID_28289718_Higgins-Patient-2</td>
      <td>{'id': 'Higgins-Patient-2', 'timeAtLastEncount...</td>
      <td>[{'type': {'id': 'HP:0001714', 'label': 'Ventr...</td>
      <td>[{'id': 'Higgins-Patient-2', 'progressStatus':...</td>
      <td>[{'term': {'id': 'OMIM:618499', 'label': 'Noon...</td>
      <td>{'created': '2024-03-28T11:11:48.592718124Z', ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>As expected, there are three rows with the OMIM disease 618499.</p>
</section>
<section id="Query-faceting">
<h2>Query faceting<a class="headerlink" href="#Query-faceting" title="Link to this heading"></a></h2>
<p>We will now demonstrate faceted queries, allowing us to count the number of instances of different categorical values or categorical value combinations.</p>
<p>First we’ll facet on the subject sex. We can use path notation, e.g. <code class="docutils literal notranslate"><span class="pre">subject.sex</span></code> here:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">query_facets</span><span class="p">({},</span> <span class="n">facet_columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;subject.sex&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;subject.sex&#39;: [(&#39;MALE&#39;, 1807), (&#39;FEMALE&#39;, 1564)]}
</pre></div></div>
</div>
<p>We can also facet by the disease name/label. We’ll restrict this to the top 20</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">query_facets</span><span class="p">({},</span> <span class="n">facet_columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;diseases.term.label&quot;</span><span class="p">],</span> <span class="n">facet_limit</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;diseases.term.label&#39;: [(&#39;Developmental and epileptic encephalopathy 4&#39;, 463),
  (&#39;Developmental and epileptic encephalopathy 11&#39;, 342),
  (&#39;KBG syndrome&#39;, 337),
  (&#39;Leber congenital amaurosis 6&#39;, 191),
  (&#39;Glass syndrome&#39;, 158),
  (&#39;Holt-Oram syndrome&#39;, 103),
  (&#39;Mitochondrial DNA depletion syndrome 13 (encephalomyopathic type)&#39;, 95),
  (&#39;Neurodevelopmental disorder with coarse facies and mild distal skeletal abnormalities&#39;,
   73),
  (&#39;Jacobsen syndrome&#39;, 69),
  (&#39;Coffin-Siris syndrome 8&#39;, 65),
  (&#39;Kabuki Syndrome 1&#39;, 65),
  (&#39;Houge-Janssen syndrome 2&#39;, 60),
  (&#39;ZTTK SYNDROME&#39;, 52),
  (&#39;Greig cephalopolysyndactyly syndrome&#39;, 51),
  (&#39;Seizures, benign familial infantile, 3&#39;, 51),
  (&#39;Mitochondrial DNA depletion syndrome 6 (hepatocerebral type)&#39;, 50),
  (&#39;Marfan syndrome&#39;, 50),
  (&#39;Developmental delay, dysmorphic facies, and brain anomalies&#39;, 49),
  (&#39;Loeys-Dietz syndrome 3&#39;, 49),
  (&#39;Hypomagnesemia 3, renal&#39;, 46)]}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">query_facets</span><span class="p">({},</span> <span class="n">facet_columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;subject.timeAtLastEncounter.age.iso8601duration&quot;</span><span class="p">],</span> <span class="n">facet_limit</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;subject.timeAtLastEncounter.age.iso8601duration&#39;: [(&#39;P4Y&#39;, 131),
  (&#39;P3Y&#39;, 114),
  (&#39;P6Y&#39;, 100),
  (&#39;P5Y&#39;, 97),
  (&#39;P2Y&#39;, 95),
  (&#39;P7Y&#39;, 85),
  (&#39;P10Y&#39;, 82),
  (&#39;P9Y&#39;, 77),
  (&#39;P8Y&#39;, 71)]}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">query_facets</span><span class="p">({},</span> <span class="n">facet_columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;interpretations.diagnosis.genomicInterpretations.variantInterpretation.variationDescriptor.geneContext.symbol&quot;</span><span class="p">],</span> <span class="n">facet_limit</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;interpretations.diagnosis.genomicInterpretations.variantInterpretation.variationDescriptor.geneContext.symbol&#39;: [(&#39;STXBP1&#39;,
   463),
  (&#39;SCN2A&#39;, 393),
  (&#39;ANKRD11&#39;, 337),
  (&#39;RPGRIP1&#39;, 273),
  (&#39;SATB2&#39;, 158),
  (&#39;FBN1&#39;, 151),
  (&#39;LMNA&#39;, 127),
  (&#39;FBXL4&#39;, 117),
  (&#39;TBX5&#39;, 103),
  (&#39;SPTAN1&#39;, 85)]}
</pre></div></div>
</div>
<p>We can also facet on combinations:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">fqr</span> <span class="o">=</span> <span class="n">collection</span><span class="o">.</span><span class="n">query_facets</span><span class="p">({},</span> <span class="n">facet_columns</span><span class="o">=</span><span class="p">[(</span><span class="s2">&quot;subject.sex&quot;</span><span class="p">,</span> <span class="s2">&quot;diseases.term.label&quot;</span><span class="p">)],</span> <span class="n">facet_limit</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">fqr</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{(&#39;subject.sex&#39;, &#39;diseases.term.label&#39;): [((&#39;MALE&#39;, &#39;KBG syndrome&#39;), 175),
  ((&#39;FEMALE&#39;, &#39;KBG syndrome&#39;), 143),
  ((&#39;MALE&#39;, &#39;Glass syndrome&#39;), 90),
  ((&#39;FEMALE&#39;, &#39;Glass syndrome&#39;), 62),
  ((&#39;MALE&#39;,
    &#39;Mitochondrial DNA depletion syndrome 13 (encephalomyopathic type)&#39;),
   58),
  ((&#39;MALE&#39;,
    &#39;Neurodevelopmental disorder with coarse facies and mild distal skeletal abnormalities&#39;),
   54),
  ((&#39;FEMALE&#39;, &#39;Jacobsen syndrome&#39;), 49),
  ((&#39;MALE&#39;, &#39;Coffin-Siris syndrome 8&#39;), 37),
  ((&#39;FEMALE&#39;,
    &#39;Mitochondrial DNA depletion syndrome 13 (encephalomyopathic type)&#39;),
   37),
  ((&#39;FEMALE&#39;, &#39;Kabuki Syndrome 1&#39;), 35),
  ((&#39;MALE&#39;, &#39;Houge-Janssen syndrome 2&#39;), 32),
  ((&#39;MALE&#39;, &#39;Kabuki Syndrome 1&#39;), 30),
  ((&#39;FEMALE&#39;, &#39;Developmental delay, dysmorphic facies, and brain anomalies&#39;),
   29),
  ((&#39;FEMALE&#39;, &#39;Holt-Oram syndrome&#39;), 28),
  ((&#39;MALE&#39;, &#39;Intellectual developmental disorder, autosomal dominant 21&#39;), 28),
  ((&#39;MALE&#39;, &#39;Cardiac, facial, and digital anomalies with developmental delay&#39;),
   28),
  ((&#39;FEMALE&#39;, &#39;Developmental and epileptic encephalopathy 28&#39;), 27),
  ((&#39;MALE&#39;, &#39;Loeys-Dietz syndrome 3&#39;), 27),
  ((&#39;MALE&#39;, &#39;ZTTK SYNDROME&#39;), 26),
  ((&#39;FEMALE&#39;, &#39;ZTTK SYNDROME&#39;), 26)]}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">linkml_store.utils.pandas_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">facet_summary_to_dataframe_unmelted</span>

<span class="n">facet_summary_to_dataframe_unmelted</span><span class="p">(</span><span class="n">fqr</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject.sex</th>
      <th>diseases.term.label</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>MALE</td>
      <td>KBG syndrome</td>
      <td>175</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FEMALE</td>
      <td>KBG syndrome</td>
      <td>143</td>
    </tr>
    <tr>
      <th>2</th>
      <td>MALE</td>
      <td>Glass syndrome</td>
      <td>90</td>
    </tr>
    <tr>
      <th>3</th>
      <td>FEMALE</td>
      <td>Glass syndrome</td>
      <td>62</td>
    </tr>
    <tr>
      <th>4</th>
      <td>MALE</td>
      <td>Mitochondrial DNA depletion syndrome 13 (encep...</td>
      <td>58</td>
    </tr>
    <tr>
      <th>5</th>
      <td>MALE</td>
      <td>Neurodevelopmental disorder with coarse facies...</td>
      <td>54</td>
    </tr>
    <tr>
      <th>6</th>
      <td>FEMALE</td>
      <td>Jacobsen syndrome</td>
      <td>49</td>
    </tr>
    <tr>
      <th>7</th>
      <td>MALE</td>
      <td>Coffin-Siris syndrome 8</td>
      <td>37</td>
    </tr>
    <tr>
      <th>8</th>
      <td>FEMALE</td>
      <td>Mitochondrial DNA depletion syndrome 13 (encep...</td>
      <td>37</td>
    </tr>
    <tr>
      <th>9</th>
      <td>FEMALE</td>
      <td>Kabuki Syndrome 1</td>
      <td>35</td>
    </tr>
    <tr>
      <th>10</th>
      <td>MALE</td>
      <td>Houge-Janssen syndrome 2</td>
      <td>32</td>
    </tr>
    <tr>
      <th>11</th>
      <td>MALE</td>
      <td>Kabuki Syndrome 1</td>
      <td>30</td>
    </tr>
    <tr>
      <th>12</th>
      <td>FEMALE</td>
      <td>Developmental delay, dysmorphic facies, and br...</td>
      <td>29</td>
    </tr>
    <tr>
      <th>13</th>
      <td>FEMALE</td>
      <td>Holt-Oram syndrome</td>
      <td>28</td>
    </tr>
    <tr>
      <th>14</th>
      <td>MALE</td>
      <td>Intellectual developmental disorder, autosomal...</td>
      <td>28</td>
    </tr>
    <tr>
      <th>15</th>
      <td>MALE</td>
      <td>Cardiac, facial, and digital anomalies with de...</td>
      <td>28</td>
    </tr>
    <tr>
      <th>16</th>
      <td>FEMALE</td>
      <td>Developmental and epileptic encephalopathy 28</td>
      <td>27</td>
    </tr>
    <tr>
      <th>17</th>
      <td>MALE</td>
      <td>Loeys-Dietz syndrome 3</td>
      <td>27</td>
    </tr>
    <tr>
      <th>18</th>
      <td>MALE</td>
      <td>ZTTK SYNDROME</td>
      <td>26</td>
    </tr>
    <tr>
      <th>19</th>
      <td>FEMALE</td>
      <td>ZTTK SYNDROME</td>
      <td>26</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Semantic-Search">
<h2>Semantic Search<a class="headerlink" href="#Semantic-Search" title="Link to this heading"></a></h2>
<p>We will index phenopackets using a template that extracts the subject, phenotypic features and diseases.</p>
<p>First we will create a textualization template for a phenopacket. We will keep it minimal for simplicity - this doesn’t include treatments, families, etc.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">template</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">subject: {{subject}}</span>
<span class="s2">phenotypes: {</span><span class="si">% f</span><span class="s2">or p in phenotypicFeatures %}{{p.type.label}}{</span><span class="si">% e</span><span class="s2">ndfor %}</span>
<span class="s2">diseases: {</span><span class="si">% f</span><span class="s2">or d in diseases %}{{d.term.label}}{</span><span class="si">% e</span><span class="s2">ndfor %}</span>
<span class="s2">&quot;&quot;&quot;</span>
</pre></div>
</div>
</div>
<p>Next we will create an indexer using the template. This will use the Jinja2 syntax for templating. We will also cache LLM embedding queries, so if we want to incrementally add new phenopackets we can avoid re-running the LLM embeddings calls.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">linkml_store.index.implementations.llm_indexer</span><span class="w"> </span><span class="kn">import</span> <span class="n">LLMIndexer</span>

<span class="n">index</span> <span class="o">=</span> <span class="n">LLMIndexer</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;ppkt&quot;</span><span class="p">,</span>
    <span class="n">cached_embeddings_database</span><span class="o">=</span><span class="s2">&quot;tmp/llm_pheno_cache.db&quot;</span><span class="p">,</span>
    <span class="n">text_template</span><span class="o">=</span><span class="n">template</span><span class="p">,</span>
    <span class="n">text_template_syntax</span><span class="o">=</span><span class="s2">&quot;jinja2&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>We can test the template on the first row of the collection:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">index</span><span class="o">.</span><span class="n">object_to_text</span><span class="p">(</span><span class="n">qr</span><span class="o">.</span><span class="n">rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

subject: {&#39;id&#39;: &#39;Higgins-Patient-1&#39;, &#39;timeAtLastEncounter&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P17Y&#39;}}, &#39;sex&#39;: &#39;FEMALE&#39;}
phenotypes: Ventricular hypertrophyHeart murmurHypertrophic cardiomyopathyShort statureHypertelorismLow-set earsPosteriorly rotated earsGlobal developmental delayCognitive impairmentCardiac arrest
diseases: Noonan syndrome-11
</pre></div></div>
</div>
<p>That looks as expected. We can now attach the indexer to the collection and index the collection:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">attach_indexer</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">auto_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="id1">
<h2>Semantic Search<a class="headerlink" href="#id1" title="Link to this heading"></a></h2>
<p>Let’s query based on text criteria:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">qr</span> <span class="o">=</span> <span class="n">collection</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;patients with liver diseases&quot;</span><span class="p">)</span>
<span class="n">qr</span><span class="o">.</span><span class="n">rows_dataframe</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>score</th>
      <th>id</th>
      <th>subject</th>
      <th>phenotypicFeatures</th>
      <th>interpretations</th>
      <th>diseases</th>
      <th>metaData</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.824639</td>
      <td>PMID_30658709_patient</td>
      <td>{'id': 'patient', 'timeAtLastEncounter': {'age...</td>
      <td>[{'type': {'id': 'HP:0031956', 'label': 'Eleva...</td>
      <td>[{'id': 'patient', 'progressStatus': 'SOLVED',...</td>
      <td>[{'term': {'id': 'OMIM:615878', 'label': 'Chol...</td>
      <td>{'created': '2024-05-05T09:03:25.388371944Z', ...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.824639</td>
      <td>PMID_30658709_patient</td>
      <td>{'id': 'patient', 'timeAtLastEncounter': {'age...</td>
      <td>[{'type': {'id': 'HP:0031956', 'label': 'Eleva...</td>
      <td>[{'id': 'patient', 'progressStatus': 'SOLVED',...</td>
      <td>[{'term': {'id': 'OMIM:615878', 'label': 'Chol...</td>
      <td>{'created': '2024-05-05T09:03:25.388371944Z', ...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.813770</td>
      <td>PMID_36932076_Patient_1</td>
      <td>{'id': 'Patient 1', 'timeAtLastEncounter': {'a...</td>
      <td>[{'type': {'id': 'HP:0000979', 'label': 'Purpu...</td>
      <td>[{'id': 'Patient 1', 'progressStatus': 'SOLVED...</td>
      <td>[{'term': {'id': 'OMIM:620376', 'label': 'Auto...</td>
      <td>{'created': '2024-04-19T06:07:57.188061952Z', ...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.813770</td>
      <td>PMID_36932076_Patient_1</td>
      <td>{'id': 'Patient 1', 'timeAtLastEncounter': {'a...</td>
      <td>[{'type': {'id': 'HP:0000979', 'label': 'Purpu...</td>
      <td>[{'id': 'Patient 1', 'progressStatus': 'SOLVED...</td>
      <td>[{'term': {'id': 'OMIM:620376', 'label': 'Auto...</td>
      <td>{'created': '2024-04-19T06:07:57.188061952Z', ...</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.804126</td>
      <td>PMID_37303127_6</td>
      <td>{'id': '6', 'timeAtLastEncounter': {'age': {'i...</td>
      <td>[{'type': {'id': 'HP:0001397', 'label': 'Hepat...</td>
      <td>[{'id': '6', 'progressStatus': 'SOLVED', 'diag...</td>
      <td>[{'term': {'id': 'OMIM:151660', 'label': 'Lipo...</td>
      <td>{'created': '2024-03-23T17:41:42.999521017Z', ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Let’s check the first one</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">qr</span><span class="o">.</span><span class="n">ranked_rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(0.824638728366563,
 {&#39;id&#39;: &#39;PMID_30658709_patient&#39;,
  &#39;subject&#39;: {&#39;id&#39;: &#39;patient&#39;,
   &#39;timeAtLastEncounter&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P1Y11M&#39;}},
   &#39;sex&#39;: &#39;FEMALE&#39;},
  &#39;phenotypicFeatures&#39;: [{&#39;type&#39;: {&#39;id&#39;: &#39;HP:0031956&#39;,
     &#39;label&#39;: &#39;Elevated circulating aspartate aminotransferase concentration&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P1Y11M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0031964&#39;,
     &#39;label&#39;: &#39;Elevated circulating alanine aminotransferase concentration&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P1Y11M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0003573&#39;, &#39;label&#39;: &#39;Increased total bilirubin&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0012202&#39;,
     &#39;label&#39;: &#39;Increased serum bile acid concentration&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0002908&#39;, &#39;label&#39;: &#39;Conjugated hyperbilirubinemia&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0001433&#39;, &#39;label&#39;: &#39;Hepatosplenomegaly&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0001510&#39;, &#39;label&#39;: &#39;Growth delay&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0000989&#39;, &#39;label&#39;: &#39;Pruritus&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0000952&#39;, &#39;label&#39;: &#39;Jaundice&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0100810&#39;, &#39;label&#39;: &#39;Pointed helix&#39;},
    &#39;onset&#39;: {&#39;age&#39;: {&#39;iso8601duration&#39;: &#39;P6M&#39;}}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0002650&#39;, &#39;label&#39;: &#39;Scoliosis&#39;}},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0003112&#39;,
     &#39;label&#39;: &#39;Abnormal circulating amino acid concentration&#39;},
    &#39;excluded&#39;: True},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0001928&#39;, &#39;label&#39;: &#39;Abnormality of coagulation&#39;},
    &#39;excluded&#39;: True},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0010701&#39;, &#39;label&#39;: &#39;Abnormal immunoglobulin level&#39;},
    &#39;excluded&#39;: True},
   {&#39;type&#39;: {&#39;id&#39;: &#39;HP:0001627&#39;, &#39;label&#39;: &#39;Abnormal heart morphology&#39;},
    &#39;excluded&#39;: True}],
  &#39;interpretations&#39;: [{&#39;id&#39;: &#39;patient&#39;,
    &#39;progressStatus&#39;: &#39;SOLVED&#39;,
    &#39;diagnosis&#39;: {&#39;disease&#39;: {&#39;id&#39;: &#39;OMIM:615878&#39;,
      &#39;label&#39;: &#39;Cholestasis, progressive familial intrahepatic 4&#39;},
     &#39;genomicInterpretations&#39;: [{&#39;subjectOrBiosampleId&#39;: &#39;patient&#39;,
       &#39;interpretationStatus&#39;: &#39;CAUSATIVE&#39;,
       &#39;variantInterpretation&#39;: {&#39;variationDescriptor&#39;: {&#39;id&#39;: &#39;var_kKNGnjOxGXMbcoWzDGEJKVPIB&#39;,
         &#39;geneContext&#39;: {&#39;valueId&#39;: &#39;HGNC:11828&#39;, &#39;symbol&#39;: &#39;TJP2&#39;},
         &#39;expressions&#39;: [{&#39;syntax&#39;: &#39;hgvs.c&#39;,
           &#39;value&#39;: &#39;NM_004817.4:c.2355+1G&gt;C&#39;},
          {&#39;syntax&#39;: &#39;hgvs.g&#39;, &#39;value&#39;: &#39;NC_000009.12:g.69238790G&gt;C&#39;}],
         &#39;vcfRecord&#39;: {&#39;genomeAssembly&#39;: &#39;hg38&#39;,
          &#39;chrom&#39;: &#39;chr9&#39;,
          &#39;pos&#39;: &#39;69238790&#39;,
          &#39;ref&#39;: &#39;G&#39;,
          &#39;alt&#39;: &#39;C&#39;},
         &#39;moleculeContext&#39;: &#39;genomic&#39;,
         &#39;allelicState&#39;: {&#39;id&#39;: &#39;GENO:0000136&#39;, &#39;label&#39;: &#39;homozygous&#39;}}}}]}}],
  &#39;diseases&#39;: [{&#39;term&#39;: {&#39;id&#39;: &#39;OMIM:615878&#39;,
     &#39;label&#39;: &#39;Cholestasis, progressive familial intrahepatic 4&#39;},
    &#39;onset&#39;: {&#39;ontologyClass&#39;: {&#39;id&#39;: &#39;HP:0003593&#39;,
      &#39;label&#39;: &#39;Infantile onset&#39;}}}],
  &#39;metaData&#39;: {&#39;created&#39;: &#39;2024-05-05T09:03:25.388371944Z&#39;,
   &#39;createdBy&#39;: &#39;ORCID:0000-0002-0736-9199&#39;,
   &#39;resources&#39;: [{&#39;id&#39;: &#39;geno&#39;,
     &#39;name&#39;: &#39;Genotype Ontology&#39;,
     &#39;url&#39;: &#39;http://purl.obolibrary.org/obo/geno.owl&#39;,
     &#39;version&#39;: &#39;2022-03-05&#39;,
     &#39;namespacePrefix&#39;: &#39;GENO&#39;,
     &#39;iriPrefix&#39;: &#39;http://purl.obolibrary.org/obo/GENO_&#39;},
    {&#39;id&#39;: &#39;hgnc&#39;,
     &#39;name&#39;: &#39;HUGO Gene Nomenclature Committee&#39;,
     &#39;url&#39;: &#39;https://www.genenames.org&#39;,
     &#39;version&#39;: &#39;06/01/23&#39;,
     &#39;namespacePrefix&#39;: &#39;HGNC&#39;,
     &#39;iriPrefix&#39;: &#39;https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/&#39;},
    {&#39;id&#39;: &#39;omim&#39;,
     &#39;name&#39;: &#39;An Online Catalog of Human Genes and Genetic Disorders&#39;,
     &#39;url&#39;: &#39;https://www.omim.org&#39;,
     &#39;version&#39;: &#39;January 4, 2023&#39;,
     &#39;namespacePrefix&#39;: &#39;OMIM&#39;,
     &#39;iriPrefix&#39;: &#39;https://www.omim.org/entry/&#39;},
    {&#39;id&#39;: &#39;so&#39;,
     &#39;name&#39;: &#39;Sequence types and features ontology&#39;,
     &#39;url&#39;: &#39;http://purl.obolibrary.org/obo/so.obo&#39;,
     &#39;version&#39;: &#39;2021-11-22&#39;,
     &#39;namespacePrefix&#39;: &#39;SO&#39;,
     &#39;iriPrefix&#39;: &#39;http://purl.obolibrary.org/obo/SO_&#39;},
    {&#39;id&#39;: &#39;hp&#39;,
     &#39;name&#39;: &#39;human phenotype ontology&#39;,
     &#39;url&#39;: &#39;http://purl.obolibrary.org/obo/hp.owl&#39;,
     &#39;version&#39;: &#39;2024-04-26&#39;,
     &#39;namespacePrefix&#39;: &#39;HP&#39;,
     &#39;iriPrefix&#39;: &#39;http://purl.obolibrary.org/obo/HP_&#39;}],
   &#39;phenopacketSchemaVersion&#39;: &#39;2.0&#39;,
   &#39;externalReferences&#39;: [{&#39;id&#39;: &#39;PMID:30658709&#39;,
     &#39;reference&#39;: &#39;https://pubmed.ncbi.nlm.nih.gov/30658709&#39;,
     &#39;description&#39;: &#39;Novel compound heterozygote mutations of TJP2 in a Chinese child with progressive cholestatic liver disease&#39;}]}})
</pre></div></div>
</div>
<p>We can combine semantic search with queries:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">qr</span> <span class="o">=</span> <span class="n">collection</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;patients with liver diseases&quot;</span><span class="p">,</span> <span class="n">where</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;subject.sex&quot;</span><span class="p">:</span> <span class="s2">&quot;MALE&quot;</span><span class="p">})</span>
<span class="n">qr</span><span class="o">.</span><span class="n">rows_dataframe</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>score</th>
      <th>id</th>
      <th>subject</th>
      <th>phenotypicFeatures</th>
      <th>interpretations</th>
      <th>diseases</th>
      <th>metaData</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.813827</td>
      <td>PMID_36932076_Patient_1</td>
      <td>{'id': 'Patient 1', 'timeAtLastEncounter': {'a...</td>
      <td>[{'type': {'id': 'HP:0000979', 'label': 'Purpu...</td>
      <td>[{'id': 'Patient 1', 'progressStatus': 'SOLVED...</td>
      <td>[{'term': {'id': 'OMIM:620376', 'label': 'Auto...</td>
      <td>{'created': '2024-04-19T06:07:57.188061952Z', ...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.813827</td>
      <td>PMID_36932076_Patient_1</td>
      <td>{'id': 'Patient 1', 'timeAtLastEncounter': {'a...</td>
      <td>[{'type': {'id': 'HP:0000979', 'label': 'Purpu...</td>
      <td>[{'id': 'Patient 1', 'progressStatus': 'SOLVED...</td>
      <td>[{'term': {'id': 'OMIM:620376', 'label': 'Auto...</td>
      <td>{'created': '2024-04-19T06:07:57.188061952Z', ...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.799738</td>
      <td>PMID_36932076_Patient_3</td>
      <td>{'id': 'Patient 3', 'timeAtLastEncounter': {'a...</td>
      <td>[{'type': {'id': 'HP:0001511', 'label': 'Intra...</td>
      <td>[{'id': 'Patient 3', 'progressStatus': 'SOLVED...</td>
      <td>[{'term': {'id': 'OMIM:620376', 'label': 'Auto...</td>
      <td>{'created': '2024-04-19T06:07:57.190312862Z', ...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.799738</td>
      <td>PMID_36932076_Patient_3</td>
      <td>{'id': 'Patient 3', 'timeAtLastEncounter': {'a...</td>
      <td>[{'type': {'id': 'HP:0001511', 'label': 'Intra...</td>
      <td>[{'id': 'Patient 3', 'progressStatus': 'SOLVED...</td>
      <td>[{'term': {'id': 'OMIM:620376', 'label': 'Auto...</td>
      <td>{'created': '2024-04-19T06:07:57.190312862Z', ...</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.799243</td>
      <td>PMID_27536553_27536553_P3</td>
      <td>{'id': '27536553_P3', 'timeAtLastEncounter': {...</td>
      <td>[{'type': {'id': 'HP:0001396', 'label': 'Chole...</td>
      <td>[{'id': '27536553_P3', 'progressStatus': 'SOLV...</td>
      <td>[{'term': {'id': 'OMIM:256810', 'label': 'Mito...</td>
      <td>{'created': '2024-03-23T19:28:35.688389062Z', ...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Validation">
<h2>Validation<a class="headerlink" href="#Validation" title="Link to this heading"></a></h2>
<p>Next we will demonstrate validation over a whole collection.</p>
<p>Currently validating depends on a LinkML schema - we have previously copied this schema into the test folder. We will load the schema into the database object:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">db</span><span class="o">.</span><span class="n">load_schema_view</span><span class="p">(</span><span class="s2">&quot;../../tests/input/schemas/phenopackets_linkml/phenopackets.yaml&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Quick sanity check to ensure that worked:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="nb">list</span><span class="p">(</span><span class="n">db</span><span class="o">.</span><span class="n">schema_view</span><span class="o">.</span><span class="n">all_classes</span><span class="p">())[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;Age&#39;,
 &#39;AgeRange&#39;,
 &#39;Dictionary&#39;,
 &#39;Evidence&#39;,
 &#39;ExternalReference&#39;,
 &#39;File&#39;,
 &#39;GestationalAge&#39;,
 &#39;OntologyClass&#39;,
 &#39;Procedure&#39;,
 &#39;TimeElement&#39;]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">collection</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;Phenopacket&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">linkml_runtime.dumpers</span><span class="w"> </span><span class="kn">import</span> <span class="n">yaml_dumper</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">iter_validate_database</span><span class="p">():</span>
    <span class="c1"># known issue - https://github.com/monarch-initiative/phenopacket-store/issues/97</span>
    <span class="k">if</span> <span class="s2">&quot;is not of type &#39;integer&#39;&quot;</span> <span class="ow">in</span> <span class="n">r</span><span class="o">.</span><span class="n">message</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">message</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">100</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unexpected validation error&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Command-Line-Usage">
<h2>Command Line Usage<a class="headerlink" href="#Command-Line-Usage" title="Link to this heading"></a></h2>
<p>We can also use the command line for all of the above operations.</p>
<p>For example, feceted queries:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>linkml-store<span class="w"> </span>-d<span class="w"> </span>mongodb://localhost:27017<span class="w"> </span>-c<span class="w"> </span>main<span class="w"> </span>fq<span class="w"> </span>-S<span class="w"> </span>subject.sex
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{
  &#34;subject.sex&#34;: {
    &#34;MALE&#34;: 1807,
    &#34;FEMALE&#34;: 1564
  }
}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>linkml-store<span class="w"> </span>-d<span class="w"> </span>mongodb://localhost:27017<span class="w"> </span>-c<span class="w"> </span>main<span class="w"> </span>fq<span class="w"> </span>-S<span class="w"> </span>phenotypicFeatures.type.label<span class="w"> </span>-O<span class="w"> </span>yaml
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
phenotypicFeatures.type.label:
  Global developmental delay: 1705
  Hypotonia: 1056
  Intellectual disability: 1028
  Seizure: 950
  Hypertelorism: 925
  Delayed speech and language development: 829
  Short stature: 806
  Microcephaly: 780
  Scoliosis: 702
  Feeding difficulties: 678
  Low-set ears: 598
  Autistic behavior: 519
  Motor delay: 518
  Downslanted palpebral fissures: 505
  Strabismus: 504
  Long philtrum: 500
  Ptosis: 498
  Patent foramen ovale: 469
  Anteverted nares: 461
  Hearing impairment: 451
  Epicanthus: 447
  Ventricular septal defect: 435
  Thick eyebrow: 433
  Cleft palate: 423
  Joint hypermobility: 388
  High palate: 383
  Triangular face: 369
  Micrognathia: 364
  Posteriorly rotated ears: 350
  Failure to thrive: 345
  Prominent forehead: 343
  Thin upper lip vermilion: 338
  Sleep abnormality: 331
  Wide nasal bridge: 331
  Infantile spasms: 325
  Long eyelashes: 325
  Pectus excavatum: 322
  Ataxia: 319
  Pes planus: 315
  Bilateral tonic-clonic seizure: 314
  Bulbous nose: 311
  Intellectual disability, severe: 306
  Nystagmus: 298
  Absent speech: 294
  Midface retrusion: 290
  Bicuspid aortic valve: 288
  Deeply set eye: 283
  Delayed ability to walk: 282
  Pulmonic stenosis: 280
  Cryptorchidism: 279
  Talipes equinovarus: 277
  Attention deficit hyperactivity disorder: 275
  Recurrent otitis media: 275
  Macrocephaly: 275
  Abnormality of the hand: 273
  Depressed nasal bridge: 273
  Autism: 270
  Macrodontia: 266
  Dystonia: 265
  Narrow forehead: 261
  Smooth philtrum: 249
  Microtia: 248
  Inguinal hernia: 247
  Upslanted palpebral fissure: 246
  Ventriculomegaly: 240
  Synophrys: 236
  Cerebellar atrophy: 234
  Ectopia lentis: 234
  Thin corpus callosum: 231
  EEG abnormality: 230
  Short philtrum: 226
  Arachnodactyly: 224
  Short neck: 223
  Highly arched eyebrow: 221
  Epileptic encephalopathy: 219
  Developmental regression: 218
  Generalized tonic seizure: 218
  Protruding ear: 217
  Atrial septal defect: 213
  Umbilical hernia: 213
  Cerebral atrophy: 212
  Atrioventricular canal defect: 206
  Low anterior hairline: 203
  Mitral valve prolapse: 199
  Focal impaired awareness seizure: 199
  Delayed skeletal maturation: 198
  Hypsarrhythmia: 198
  Intrauterine growth retardation: 196
  Hypoplasia of the corpus callosum: 192
  Spasticity: 192
  Growth delay: 186
  Aortic root aneurysm: 181
  Severe global developmental delay: 173
  Multifocal epileptiform discharges: 169
  Mandibular prognathia: 167
  Dysarthria: 167
  Patent ductus arteriosus: 166
  Blue sclerae: 166
  Proptosis: 164
  Cataract: 162

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>linkml-store<span class="w"> </span>-d<span class="w"> </span>mongodb://localhost:27017<span class="w"> </span>-c<span class="w"> </span>main<span class="w"> </span>fq<span class="w"> </span>-S<span class="w"> </span>diseases.term.label+subject.sex<span class="w"> </span>-O<span class="w"> </span>yaml
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
diseases.term.label+subject.sex:
  (&#39;KBG syndrome&#39;, &#39;MALE&#39;): 175
  (&#39;KBG syndrome&#39;, &#39;FEMALE&#39;): 143
  (&#39;Glass syndrome&#39;, &#39;MALE&#39;): 90
  (&#39;Glass syndrome&#39;, &#39;FEMALE&#39;): 62
  (&#39;Mitochondrial DNA depletion syndrome 13 (encephalomyopathic type)&#39;, &#39;MALE&#39;): 58
  (&#39;Neurodevelopmental disorder with coarse facies and mild distal skeletal abnormalities&#39;, &#39;MALE&#39;): 54
  (&#39;Jacobsen syndrome&#39;, &#39;FEMALE&#39;): 49
  (&#39;Coffin-Siris syndrome 8&#39;, &#39;MALE&#39;): 37
  (&#39;Mitochondrial DNA depletion syndrome 13 (encephalomyopathic type)&#39;, &#39;FEMALE&#39;): 37
  (&#39;Kabuki Syndrome 1&#39;, &#39;FEMALE&#39;): 35
  (&#39;Houge-Janssen syndrome 2&#39;, &#39;MALE&#39;): 32
  (&#39;Kabuki Syndrome 1&#39;, &#39;MALE&#39;): 30
  (&#39;Developmental delay, dysmorphic facies, and brain anomalies&#39;, &#39;FEMALE&#39;): 29
  (&#39;Intellectual developmental disorder, autosomal dominant 21&#39;, &#39;MALE&#39;): 28
  (&#39;Holt-Oram syndrome&#39;, &#39;FEMALE&#39;): 28
  (&#39;Cardiac, facial, and digital anomalies with developmental delay&#39;, &#39;MALE&#39;): 28
  (&#39;Loeys-Dietz syndrome 3&#39;, &#39;MALE&#39;): 27
  (&#39;Developmental and epileptic encephalopathy 28&#39;, &#39;FEMALE&#39;): 27
  (&#39;ZTTK SYNDROME&#39;, &#39;FEMALE&#39;): 26
  (&#39;ZTTK SYNDROME&#39;, &#39;MALE&#39;): 26
  (&#39;Loeys-Dietz syndrome 4&#39;, &#39;MALE&#39;): 26
  (&#39;Marfan syndrome&#39;, &#39;MALE&#39;): 26
  (&#39;Hypomagnesemia 3, renal&#39;, &#39;MALE&#39;): 26
  (&#39;Intellectual developmental disorder, X-linked 112&#39;, &#39;MALE&#39;): 26
  (&#39;Mitochondrial DNA depletion syndrome 6 (hepatocerebral type)&#39;, &#39;MALE&#39;): 26
  (&#39;Marfan syndrome&#39;, &#39;FEMALE&#39;): 24
  (&#39;Ectopia lentis, familial&#39;, &#39;MALE&#39;): 24
  (&#39;Coffin-Siris syndrome 8&#39;, &#39;FEMALE&#39;): 24
  (&#39;Mitochondrial DNA depletion syndrome 6 (hepatocerebral type)&#39;, &#39;FEMALE&#39;): 24
  (&#39;Houge-Janssen syndrome 2&#39;, &#39;FEMALE&#39;): 24
  (&#39;Cardiomyopathy, dilated, 1A&#39;, &#39;MALE&#39;): 23
  (&#39;Loeys-Dietz syndrome 5&#39;, &#39;MALE&#39;): 23
  (&#39;Holt-Oram syndrome&#39;, &#39;MALE&#39;): 22
  (&#39;Mitochondrial complex IV deficiency, nuclear type 2&#39;, &#39;MALE&#39;): 22
  (&#39;Loeys-Dietz syndrome 3&#39;, &#39;FEMALE&#39;): 22
  (&#39;Cardiomyopathy, dilated, 1A&#39;, &#39;FEMALE&#39;): 21
  (&#39;Kufor-Rakeb syndrome&#39;, &#39;MALE&#39;): 21
  (&#39;Jacobsen syndrome&#39;, &#39;MALE&#39;): 20
  (&#39;Developmental delay, dysmorphic facies, and brain anomalies&#39;, &#39;MALE&#39;): 20
  (&#39;Ectopia lentis, familial&#39;, &#39;FEMALE&#39;): 20
  (&#39;Ehlers-Danlos syndrome, vascular type&#39;, &#39;FEMALE&#39;): 20
  (&#39;Loeys-Dietz syndrome 5&#39;, &#39;FEMALE&#39;): 20
  (&#39;Neurodevelopmental disorder with coarse facies and mild distal skeletal abnormalities&#39;, &#39;FEMALE&#39;): 19
  (&#39;Hypomagnesemia 3, renal&#39;, &#39;FEMALE&#39;): 19
  (&#39;Intellectual developmental disorder, autosomal dominant 21&#39;, &#39;FEMALE&#39;): 18
  (&#39;Acrofacial dysostosis 1, Nager type&#39;, &#39;FEMALE&#39;): 18
  (&#39;LEOPARD syndrome 1&#39;, &#39;MALE&#39;): 18
  (&#39;Anemia, sideroblastic, and spinocerebellar ataxia&#39;, &#39;MALE&#39;): 18
  (&#39;Spastic ataxia 8, autosomal recessive, with hypomyelinating leukodystrophy&#39;, &#39;MALE&#39;): 18
  (&#39;Albinism, oculocutaneous, type IV&#39;, &#39;FEMALE&#39;): 17
  (&#39;Cardiac, facial, and digital anomalies with developmental delay&#39;, &#39;FEMALE&#39;): 17
  (&#39;Developmental and epileptic encephalopathy 28&#39;, &#39;MALE&#39;): 16
  (&#39;Developmental delay with or without epilepsy&#39;, &#39;MALE&#39;): 16
  (&#39;Aarskog-Scott syndrome&#39;, &#39;MALE&#39;): 16
  (&#39;Ehlers-Danlos syndrome, vascular type&#39;, &#39;MALE&#39;): 15
  (&#39;Spastic paraplegia 91, autosomal dominant, with or without cerebellar ataxia&#39;, &#39;FEMALE&#39;): 15
  (&#39;Spastic ataxia 8, autosomal recessive, with hypomyelinating leukodystrophy&#39;, &#39;FEMALE&#39;): 15
  (&#39;Marfan lipodystrophy syndrome&#39;, &#39;FEMALE&#39;): 15
  (&#39;Noonan syndrome 1&#39;, &#39;MALE&#39;): 14
  (&#39;Sulfite oxidase deficiency&#39;, &#39;MALE&#39;): 14
  (&#39;Spastic paraplegia 91, autosomal dominant, with or without cerebellar ataxia&#39;, &#39;MALE&#39;): 13
  (&#39;Developmental and epileptic encephalopathy 112&#39;, &#39;FEMALE&#39;): 13
  (&#39;Noonan syndrome 1&#39;, &#39;FEMALE&#39;): 13
  (&#39;Albinism, oculocutaneous, type IV&#39;, &#39;MALE&#39;): 13
  (&#39;Neurodevelopmental disorder with motor and language delay, ocular defects, and brain abnormalities&#39;, &#39;FEMALE&#39;): 13
  (&#39;Developmental and epileptic encephalopathy 5&#39;, &#39;FEMALE&#39;): 13
  (&#39;LEOPARD syndrome 1&#39;, &#39;FEMALE&#39;): 13
  (&#39;Loeys-Dietz syndrome 2&#39;, &#39;MALE&#39;): 13
  (&#39;Kufor-Rakeb syndrome&#39;, &#39;FEMALE&#39;): 12
  (&#39;Ataxia-pancytopenia syndrome&#39;, &#39;MALE&#39;): 12
  (&#39;Autoinflammatory syndrome, familial, with or without immunodeficiency&#39;, &#39;FEMALE&#39;): 12
  (&#39;Neurodevelopmental disorder with or without anomalies of the brain, eye, or heart&#39;, &#39;MALE&#39;): 12
  (&#39;Hypotonia, infantile, with psychomotor retardation and characteristic facies 3&#39;, &#39;FEMALE&#39;): 12
  (&#39;Acrofacial dysostosis, Cincinnati type&#39;, &#39;MALE&#39;): 11
  (&#39;Noonan syndrome 2&#39;, &#39;FEMALE&#39;): 11
  (&#39;Sulfite oxidase deficiency&#39;, &#39;FEMALE&#39;): 11
  (&#39;HMG-CoA synthase-2 deficiency&#39;, &#39;MALE&#39;): 11
  (&#39;Hypotonia, infantile, with psychomotor retardation and characteristic facies 3&#39;, &#39;MALE&#39;): 11
  (&#39;Neurodevelopmental disorder with or without variable brain abnormalities&#39;, &#39;MALE&#39;): 11
  (&#39;Autoimmune polyendocrinopathy syndrome , type I, with or without reversible metaphyseal dysplasia&#39;, &#39;FEMALE&#39;): 11
  (&#39;Neurodevelopmental disorder with progressive microcephaly, spasticity, and brain anomalies&#39;, &#39;MALE&#39;): 10
  (&#39;Spastic paraplegia 76, autosomal recessive&#39;, &#39;FEMALE&#39;): 10
  (&#39;Coffin-Siris syndrome 3&#39;, &#39;FEMALE&#39;): 10
  (&#39;Noonan syndrome 6&#39;, &#39;MALE&#39;): 10
  (&#39;Loeys-Dietz syndrome 6&#39;, &#39;FEMALE&#39;): 10
  (&#39;Cornelia de Lange syndrome 6&#39;, &#39;MALE&#39;): 10
  (&#39;EZH1-related neurodevelopmental disorder&#39;, &#39;FEMALE&#39;): 10
  (&#39;Multiple mitochondrial dysfunctions syndrome 4&#39;, &#39;FEMALE&#39;): 9
  (&#39;Intellectual developmental disorder, autosomal dominant 70&#39;, &#39;MALE&#39;): 9
  (&#39;Neurodevelopmental disorder with or without variable brain abnormalities&#39;, &#39;FEMALE&#39;): 9
  (&#39;Developmental and epileptic encephalopathy 5&#39;, &#39;MALE&#39;): 9
  (&#39;Distal renal tubular acidosis 1&#39;, &#39;FEMALE&#39;): 9
  (&#39;Developmental and epileptic encephalopathy 112&#39;, &#39;MALE&#39;): 9
  (&#39;Noonan syndrome 2&#39;, &#39;MALE&#39;): 9
  (&#39;Parkinson disease 15, autosomal recessive&#39;, &#39;MALE&#39;): 9
  (&#39;Ataxia-pancytopenia syndrome&#39;, &#39;FEMALE&#39;): 9
  (&#39;Muscular dystrophy, limb-girdle, autosomal recessive 28&#39;, &#39;MALE&#39;): 9
  (&#39;Immunoskeletal dysplasia with neurodevelopmental abnormalitie&#39;, &#39;FEMALE&#39;): 9
  (&#39;Joubert syndrome 10&#39;, &#39;MALE&#39;): 9
  (&#39;Contractural arachnodactyly, congenital&#39;, &#39;FEMALE&#39;): 9

</pre></div></div>
</div>
</section>
<section id="Inference">
<h2>Inference<a class="headerlink" href="#Inference" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">linkml_store.inference</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_inference_engine</span>

<span class="n">predictor</span> <span class="o">=</span> <span class="n">get_inference_engine</span><span class="p">(</span><span class="s2">&quot;sklearn&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">predictor</span><span class="o">.</span><span class="n">load_and_split_data</span><span class="p">(</span><span class="n">collection</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">predictor</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">target_attributes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;diseases.term.label&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="Index-GO-CAMs.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">How to index GO-CAMs with LinkML-Store</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="Store-an-Ontology.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Example: Storing an ontology</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2025, Author 1 &lt;author@org.org&gt;
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">How to index Phenopackets with LinkML-Store</a><ul>
<li><a class="reference internal" href="#Use-pystow-to-download-phenopackets">Use pystow to download phenopackets</a></li>
<li><a class="reference internal" href="#Creating-a-client-and-attaching-to-a-database">Creating a client and attaching to a database</a></li>
<li><a class="reference internal" href="#Creating-a-collection">Creating a collection</a></li>
<li><a class="reference internal" href="#Inserting-objects-into-the-store">Inserting objects into the store</a></li>
<li><a class="reference internal" href="#Check-contents">Check contents</a></li>
<li><a class="reference internal" href="#Query-faceting">Query faceting</a></li>
<li><a class="reference internal" href="#Semantic-Search">Semantic Search</a></li>
<li><a class="reference internal" href="#id1">Semantic Search</a></li>
<li><a class="reference internal" href="#Validation">Validation</a></li>
<li><a class="reference internal" href="#Command-Line-Usage">Command Line Usage</a></li>
<li><a class="reference internal" href="#Inference">Inference</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/jquery.js?v=5d32c60e"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script src="../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>